{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1\n",
    "'''\n",
    "Let's remeber how to use dictionaries.\n",
    "Task: return a dictionary where \n",
    "    * keys are IDs of seqs from an input fasta file (prot.fasta),\n",
    "    * key's values are seqs itself. \n",
    "'''\n",
    "\n",
    "def my_own_fasta_parser(inFile):\n",
    "\n",
    "    sequences = {}\n",
    "\n",
    "    with #<<<<<<<<<<...>>>>>>>>>>>\n",
    "        for #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            line = #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            if line#<<<<<<<<<<...>>>>>>>>>>>\n",
    "                seq_id = #<<<<<<<<<<...>>>>>>>>>>>[1:]\n",
    "            else:\n",
    "                #<<<<<<<<<<...>>>>>>>>>>> = line\n",
    "\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2\n",
    "'''\n",
    "Another super easy task.\n",
    "\n",
    "We have the same fasta file (prot.fasta).\n",
    "Now we want to get a list with the ids of protein seqs that have \n",
    "a relative frequency higher than a given threshold for a given residue.\n",
    "\n",
    "And don't forget to use my_own_fasta_parser function from a previous task!\n",
    "'''\n",
    "\n",
    "def my_own_residue_abundance(input_file, residue, threshold=0.2):\n",
    "    \n",
    "    seq_id = []\n",
    "    sequences = #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "    for seq_id, sequence in sequences.items():\n",
    "        #<<<<<<<<<<...>>>>>>>>>>>\n",
    "        #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "    return seq_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3\n",
    "'''\n",
    "Let's practice in Bio package using.\n",
    "Task:\n",
    "1. read fasta file containing several DNA seqs (nucl.fasta)\n",
    "2. subset seqs that have GC content > 45 and coding protein with aromaticity > 0.01\n",
    "3. write a new fasta file with those protein(!) seqs\n",
    "4. return the percentage of seqs that passed your filter\n",
    "Hint: Bio.SeqIO, Bio.SeqRecord, Bio.SeqUtils\n",
    "'''\n",
    "\n",
    "import #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "def my_own_filtering(input_file, output_file, filt_gc=45, filt_arom=0.125):\n",
    "    \n",
    "    sequences = {}\n",
    "    c = 0\n",
    "    \n",
    "    with open(input_file, \"r\") as content:\n",
    "        \n",
    "        for record in SeqIO.parse(content, \"fasta\"):\n",
    "            c+=1\n",
    "            \n",
    "            # calculate GC content using Bio\n",
    "            #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            \n",
    "            # calculate aromaticity using Bio\n",
    "            #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            #<<<<<<<<<<...>>>>>>>>>>>\n",
    "            \n",
    "            # so, now you can filter\n",
    "            if calc_gc >= filt_gc and calc_arom >= filt_arom:\n",
    "                #<<<<<<<<<<...>>>>>>>>>>>\n",
    "    \n",
    "    # write a new fasta file with aminoacids\n",
    "    records = []\n",
    "    for seq_id, seq in sequences.items():\n",
    "        #<<<<<<<<<<...>>>>>>>>>>>\n",
    "    #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "    # print the percentage\n",
    "    print#<<<<<<<<<<...>>>>>>>>>>>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4\n",
    "\"\"\"\n",
    "Continue practicing in Bio package using:\n",
    "Task:\n",
    "complete the following code that should be able to return \n",
    "the best alignment of two amino acid seqs (pairwise2.align.globalds)\n",
    "based on BLOSUM62 matrix from Bio.SubsMat.MatrixInfo.\n",
    "http://rosalind.info/glossary/blosum62/\n",
    "\"\"\"\n",
    "\n",
    "import #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "def balign(first_seq, second_seq, #<<<<<<<<<<...>>>>>>>>>>>):\n",
    "\n",
    "    # Load the matrix\n",
    "    matrix = matlist.blosum62\n",
    "\n",
    "    # Generate the alignments\n",
    "    alns = pairwise2#<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "    # Extract the best alignment (first one in the alns list)\n",
    "    top_aln = #<<<<<<<<<<...>>>>>>>>>>>\n",
    "\n",
    "    # Print the alignment, ...\n",
    "    aln_A, aln_B, score, begin, end = #<<<<<<<<<<...>>>>>>>>>>>\n",
    "    #<<<<<<<<<<...>>>>>>>>>>>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5\n",
    "\"\"\" You have some DNA sequence: AGTACTAGAGCATTCTATGGAG.\n",
    "Find out which peptides could be created from it and sort them by their length.\n",
    "Use as much Biopython modules as possible.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6\n",
    "\"\"\" TASK: Try to create one-line function (without (!!!) using Bio package) \n",
    "that returns reverse complementary to a given sequence. \n",
    "Hint: using dictionaty & list comprehensions might be helpful.\n",
    "\"\"\"\n",
    "\n",
    "def rev_compl_one_line(seq):\n",
    "    return #<<<<<<<<<<...>>>>>>>>>>>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
